proposals/0000-existentials.rst

```diff
@@ -0,0 +1,504 @@
+          = GReflEC CoreExpr CoercionR
+            -- typing rule:
+            -- Γ ⊢ expr |> co : ty
+            -- ------------------------------------------
+            -- Γ ⊢ GReflEC expr co : expr ~ (expr |> co)
+
+          | TransformEC CoreExpr CoreExpr String
+            -- typing rule:
+            -- Γ ⊢ e1 : ty
+            -- Γ ⊢ e2 : ty
+            -- --------------------------------------
+            -- Γ ⊢ TransformEC e1 e2 reason : e1 ~ e2
+
+      The ``TransformEC`` form allows us to create coercions witnessing the equality of
+      any two expressions. The idea is that we use this when we need to prove that an
+      optimization GHC performs is sound. We put the name of the optimization in the
+      carried string. Perhaps in the future, this will become more principled.
+
+   #. The ``InstCo`` and ``NthCo`` coercion forms now work on ``ExistsTy`` analogously
+      to how they work on ``ForAllTy``.
+
+   #. Suppose ``f args :: C /\ ty`` and the constraint ``C`` is used. GHC will then
+      generate bindings that look like ::
+
+        let result :: C /\ ty
+            result = f args
+
+            dictC :: C
+            dictC = fstC result   -- fstC :: forall c ty. c /\ ty -> c
+
+      in its evidence bindings. Note the separate binding for ``result``. This will
+      mean that multiple uses of ``f args`` in the body of a function will get commoned
+      up during optimization. This is important in order to avoid unexpected repeated
+      evaluation of ``f args`` due to the use of ``C``.
+
+   .. _Ambiguity:
+
+#. **Ambiguity.** When a function call ``f args`` returns a dictionary (with ``/\``),
+   any use of that dictionary will require evaluating ``f args``. If multiple such
+   expressions return dictionaries of the same type (and the dictionary gets used),
+   it is now unclear which expression to evaluate. Here is a contrived example::
+
+     manufacture :: forall a. Int -> Show a /\ ()
+     manufacture = ...
+
+     ambig :: forall a. a -> String
+     ambig x = let () = manufacture @a 1
+                   () = manufacture @a 2
+               in show x
+
+   The use of ``show x`` requires a ``Show a`` dictionary, but there are two possible
+   such dictionaries: one in the result of ``manufacture @a 1`` and the other in the
+   result of ``manufacture @a 2``. Even if these dictionaries are computationally
+   identical (that is, coherent), the two calls to ``manufacture`` might have different
+   side effects (such as running time or exceptions thrown). Interestingly, neither call to
+   ``manufacture`` would be evaluated without the use of the dictionary, because ``let``
+   is lazy.
+
+   Accordingly, if there are multiple givens arising from ``/\`` that could solve a wanted,
+   we choose none of them. If there are no other ways of solving the wanted, then the
+   user's program is rejected. Alternatively, we could work this new flavor of given
```

Comment:
So how can the user eliminate the ambiguity? Let’s say I have

```
     manufacture :: forall a. Int -> Show a /\ Int
     manufacture = ...

     ambig :: forall a. a -> (String, Int)
     ambig x = let n = manufacture @a 1
                   m = manufacture @a 2
               in (show x, n + m)
```

This would be rejected due to ambiguity. So we need to export something like `discardEvidence :: c /\ a -> a` and then the user could easily disambiguate:

```diff
     ambig :: forall a. a -> (String, Int)
     ambig x = let n = manufacture @a 1
-                   m = manufacture @a 2
+                   m = discardEvidence (manufacture @a 2)
               in (show x, n + m)
```

Reply:
On second thought, `discardEvidence` is easy to define:

```
discardEvidence :: c /\ a -> a
discardEvidence a = a
```

Still, it probably belongs to `GHC.Exists`.

Reply:
The problem is that `discardEvidence` isn't actually powerful enough. According to the proposal, the given from `manufacture @a 2` is available everywhere within the innermost `case` or lambda. So `discardEvidence` doesn't, as written. Instead, I think we need

```hs
discardEvidence :: (() => c /\ a) -> a
```

with a higher-rank type. Then, there will be an invisible lambda around the `manufacture` call, which is enough to stop the given from escaping. I will add to the proposal.